src/coreclr/jit/helperexpansion.cpp

```diff
@@ -1282,59 +1282,66 @@ bool Compiler::fgVNBasedIntrinsicExpansionForCall_ReadUtf8(BasicBlock** pBlock,
         return false;
     }
 
+    assert(strObj != nullptr);
+
     // We mostly expect string literal objects here, but let's be more agile just in case
     if (!info.compCompHnd->isObjectImmutable(strObj))
     {
         JITDUMP("ReadUtf8: srcPtr is not immutable (not a frozen string object?)\n")
         return false;
     }
 
-    GenTree* srcLen = call->gtArgs.GetUserArgByIndex(1)->GetNode();
+    const GenTree* srcLen = call->gtArgs.GetUserArgByIndex(1)->GetNode();
     if (!srcLen->gtVNPair.BothEqual() || !vnStore->IsVNInt32Constant(srcLen->gtVNPair.GetLiberal()))
     {
         JITDUMP("ReadUtf8: srcLen is not constant\n")
         return false;
     }
 
-    const int      MaxPossibleUnrollThreshold = 256;
-    const unsigned unrollThreshold            = min(getUnrollThreshold(UnrollKind::Memcpy), MaxPossibleUnrollThreshold);
-    const unsigned srcLenCns                  = (unsigned)vnStore->GetConstantInt32(srcLen->gtVNPair.GetLiberal());
-    if ((srcLenCns == 0) || (srcLenCns > unrollThreshold))
+    // Source UTF16 (U16) string length in characters
+    const unsigned srcLenCnsU16            = (unsigned)vnStore->GetConstantInt32(srcLen->gtVNPair.GetLiberal());
+    const int      MaxU16BufferSizeInChars = 256;
+    if ((srcLenCnsU16 == 0) || (srcLenCnsU16 > MaxU16BufferSizeInChars))
```

Comment:
Is 256 max length a reasonable limit? If this is for string interpolation it's presumably not unreasonable to expect to see quite large string literals in some cases, so might be better to increase this threshold substantially (e.g. to 64K) and use heap allocation instead.

Reply:
Ah ignore this -- I guess the problem is that we end up with way too much inlined codegen then

Reply:
Currently, we do conversion by emitting unrolled copy for utf8 so long strings will produce big codegen. At certain lengths, I guess, UTF16->UTF8 conversion will not be 2X slower than unrolled because it's vectorized too